### PR TITLE
feat: JupyterLite integration

### LINKED ISSUE #250: Add option to run the notebooks from the website
It would be super cool if the website could allow running the notebooks directly, but that may not be as easy to implement in our current notes. However, we could add next to the "download notebook" a "run notebook" button that opens a jupyter lite notebook (or lab) and can be executed.
It would be cool if [we could deploy it within the same repository](https://jupyterlite.readthedocs.io/en/stable/quickstart/deploy.html).

This [post contains some examples of pages running it](https://blog.ouseful.info/2022/03/03/embedding-jupyterlite-in-browser-notebooks-in-documentation-and-online-educational-materials/).

### PR DESCRIPTION
**Pull request by Saransh-cpp**
Closes #250 

@dpshelio the preview for this is available on my fork - https://saransh-cpp.github.io/rsd-engineeringcourse.

Going to https://saransh-cpp.github.io/rsd-engineeringcourse/jupyter-lite will redirect the users to -

image

I couldn't embed the page within the original website + all the examples having these embedded kernels are out of date/have switched to the non-embedded format (like the one in this PR).

For adding a button adjacent to "Notebook download" button, should I make a PR to the theme repository, making changes somewhere around here -

https://github.com/dpshelio/indigo-jekyll/blob/7d12d6e04f79bbb97aa59f416746409391d0cbaf/_includes/base.html#L153

### CODE REVIEW

File: .github/workflows/build_site.yml [hunk 1]
```diff
@@ -49,6 +49,15 @@ jobs:
...
       - name: Build website
         run: bundle exec jekyll build
+      - name: Build the JupyterLite site
+        run: |
+          rm -rf `find . -type d -name .pytest_cache`
+          rm -rf `find . -type d -name .ruff_cache`
+          rm -rf `find . -type d -name .doctrees`
+          mkdir content
+          mkdir _site/jupyter-lite
+          cp -R ch*/*.ipynb content
```
Comment (dpshelio) [inline .github/workflows/build_site.yml:59]: Maybe we should keep the directories, to make it easier to browse all the files. Maybe as
```suggestion
          find ./ -wholename 'ch*/*.ipynb' -exec cp --parents {}  content \;
```

**Review by dpshelio — CHANGES_REQUESTED** [review]
Awesome! It works almost perfectly!
Some things to improve:
- keep directories structure
- `geopy` doesn't work, though it's part of the requirements. Do you know what cold be producing that?
- I'd like to get a menu next to download notebook / like `run notebook` or jupyterlite icon, but that's is in the [theme](https://github.com/dpshelio/indigo-jekyll/blob/7d12d6e04f79bbb97aa59f416746409391d0cbaf/_includes/base.html#L153) repository. We can do it after we fix the other things.

**Comment by Saransh-cpp** [comment]
> geopy doesn't work, though it's part of the requirements. Do you know what cold be producing that?

Hmm, this is weird, looking into it.

File: .github/workflows/build_site.yml [hunk 2]
```diff
@@ -49,6 +49,13 @@ jobs:
           bundle install
       - name: Build website
         run: bundle exec jekyll build
+      - name: Build the JupyterLite site
+        run: |
+          find . -type d -name .pytest_cache -o -name .ruff_cache -o -name .doctrees` | xargs rm -rf
```
Comment (Saransh-cpp) [inline .github/workflows/build_site.yml:54]: One quick CI fix
```suggestion
          find . -type d -name .pytest_cache -o -name .ruff_cache -o -name .doctrees | xargs rm -rf
```

**Comment by Saransh-cpp** [comment]
> geopy doesn't work, though it's part of the requirements. Do you know what cold be producing that?

I looked into this, and turns out jupyterlite does not really read dependencies from the existing environment. We can either [Ship additional Pyodide wheels at build time](https://jupyterlite.readthedocs.io/en/stable/howto/pyodide/wheels.html) or add a `%pip install geopy` cell at top. I talked to one of the pyodide developers and shipping additional wheels would just be an overkill for this. Also, students will have to install geopy if they download the notebook, so I think adding an installation statement in the notebooks should be good ✅

File: .gitignore [hunk 3]
```diff
@@ -87,3 +87,5 @@ session04/greetings/scripts/
 Gemfile.lock
 .env/
 polynomials.svg
+content/
```
Comment (Saransh-cpp) [inline .gitignore:90]:
```suggestion
contents/
```

**Comment by agriyakhetarpal** [comment]
Hi there, if either of you would like to add and maintain `geopy` in the Pyodide repository with a few associated tests, please feel welcome to do so. Having a maintainers for package recipes lets us know who to ping if there's a problem, since they as experienced users of said package(s) can help fix it when we bump the versions from time to time. :)

**Comment by Saransh-cpp** [comment]
Thanks for the suggestion! One thing that I was looking at (after you shared it) was [Creating a Pyodide package](https://pyodide.org/en/stable/development/new-packages.html). The guide does say -

> If you wish to use a package in Pyodide that is not already included in the [packages folder](https://github.com/pyodide/pyodide/tree/main/packages), first you need to determine whether it is necessary to package it for Pyodide. Ideally, you should start this process with package dependencies.
>
> Most pure Python packages can be installed directly from PyPI with [micropip.install()](https://micropip.pyodide.org/en/v0.2.2/project/api.html#micropip.install) if they have a pure Python wheel. Check if this is the case by trying micropip.install("package-name").

Given that geopy is a pure Python package, should it be added as a "package" in Pyodide?

**Comment by agriyakhetarpal** [comment]
Yes, to avoid the usual answer of "it depends": I think it's sometimes good to add pure Python packages to Pyodide, and many of them exist – especially those in Python's scientific computing ecosystem. For example, Zarr is a pure Python package, but it relies on async programming functionality, which means that it's good to include it and test it from time to time. I see that `geopy` depends on `geographiclib`, which from a quick search is a C++-based package, so it could be worth adding (haven't looked at that package in detail, though). Additionally, `geopy` offers a few extra features through optional dependencies, so if you're using those in your course material, they could benefit from some end-to-end testing ;)

Additing it would constrain you to use the version of it available with the Pyodide version, though, unless you disable installing it and ask for the one to be used from PyPI explicitly. If that's a bother and you would like to keep using the latest versions available on PyPI for your convenience, then you may skip doing this. Nevertheless, it would be required to add it if `geopy` were to adopt a C/C++ extension someday and switch away from being a pure Python package.

**Comment by Saransh-cpp** [comment]
Thanks for the details! geopy, in fact, is not working out of the box in jupyterlite. I have created an issue in their repository, but it looks like the development there has stalled? I'll keep a look if I get any replies from them.

**Comment by Saransh-cpp** [comment]
Copying from https://github.com/geopy/geopy/issues/578

> ... the issue is originating from pyodide's limitations and not from geopy. The other _requests_ in our course repo fail in pyodide as well, confirming that geopy is not the issue here ...

Added comments to mark bash and requests failure on the browser.

**Review by dpshelio — APPROVED** [review]
:rocket: Thank you for all this!!

**Pull request merged**